mvcc: reduce false-positive aborts

Astronomax opened the pull request:
It was discovered that Memtx MVCC aborts transactions too frequently, even in cases where it would seem possible to avoid doing so. The issue was in how potential duplicates in secondary indexes were handled when a statement became "prepared".

The processing worked as follows. When a replace/insert statement `stmt` becomes "prepared" (`stmt->add_story != NULL`):

- Iterate through each chain (except the primary one) upwards from `stmt->add_story`
- For each `story` along the path, check:
    - If the `story` was added by the same transaction as `stmt->add_story`, skip it. The check has already been performed for such stories and nothing could have been corrupted.
    - If the `story` was added by a transaction that previously deleted (**using DELETE statement**) a tuple with the same key, skip it. It couldn't have been a duplicate of `stmt->add_story->tuple` because at the serialization point when `story->tuple` was inserted, `stmt->add_story->tuple` was definitely deleted.
    - If `story->tuple` has a primary key different from the primary key of `stmt->add_story->tuple` (`story->add_stmt->del_story != stmt->add_story`), then abort `story->add_stmt->txn` - it's a duplicate.

This patch addresses the second point. Actually, it doesn't matter to us which type of statement deleted a tuple with the same key: a `DELETE` or a `REPLACE`. Up until now, we have only handled `DELETE`, because its `del_story` reference is constant and always points to the same tuple, and the read tracker "monitors" the lifetime of the tuple pointed to by this reference. Because the reference is set only once, it was easy for us to handle this case. With `REPLACE` statements, things are a bit more complicated, as their `del_story` reference can change any number of times. At different points in time, this property ("the story was added by a transaction that previously replaced a tuple with the same key") can toggle between true and false as other transactions become "prepared". Therefore, we need some mechanism to quickly handle the queries "transaction `TXN` has become prepared" and "does our property hold for story `story`?".

~To handle such cases, this patch introduces an additional hashmap named `dels` into the `tx_manager` structure: `(txn, story) -> stmt`. The presence of an entry `(txn, story) -> stmt` in this hashmap signifies that transaction `txn` deleted/replaced `story->tuple` by executing statement `stmt`. Note that within a single transaction, at most one statement could have deleted/replaced a specific `story`. Now, the second point in the duplicate check becomes very straightforward.~

Closes #11927, #11942

NO_DOC=bugfix

lenkis commented on changelogs/unreleased/gh-11660-mvcc-rollback-may-cause-duplicates.md:
```diff
@@ -0,0 +1,4 @@
+## bugfix/mvcc
+
+* Fixed an MVCC bug that could lead to duplicates in secondary indexes
+after a rollback.
```
```suggestion
* Fixed an MVCC bug that could lead to duplicates in secondary indexes
  after a rollback (gh-11927).
```

drewdzzz commented:
Awesome benchmarks!

But they measure only the worst case scenario that is unlikely to appear in the real life. What's about common case when all replaces are distributed across a large space and each chain has about 1-3 stories on average?

CuriousGeorgiy commented:
Also, what is the memory overhead?

Astronomax commented:
The memory overhead in the case of a single global hashmap is approximately:
`(x * 2 / MH_DENSITY) * (sizeof(struct dels_item) + sizeof(uint32_t) / 16) = (x * 2 / 0.7) * (32 + 4 / 16) ≈ x * 92`,
where `x` is the number of elements in the hashmap. In our case, this is the number of pairs `(txn_stmt, story)`, where `txn_stmt` deletes a `story`. A single statement can delete no more than one story, so `x` can be upper-bounded by the number of in-progress statements. This applies if no resize occurs. If a resize does occur, the estimate can be multiplied by 2.

In the case where each story has its own hashmap, if all of them are sufficiently full, the upper estimate will be fairly accurate. However, if they are all sparsely populated (1–3 elements), which is probably much more likely in real-world scenarios, the upper estimate should be multiplied by 2–3.

Astronomax commented:
@drewdzzz, @CuriousGeorgiy, thanks for the right questions! Indeed, the previous benchmarks were not very indicative. There were 2 secondary indexes (`index_count` = 3), meaning the `master` rolled back all transactions except 1. Moreover, the previous results were obtained in a debug build. :confounded:

### benchmark 1

`index_count` = 1 also wouldn't make sense for performance comparison, because `handle_dups_in_secondary_indexes`, which looks up a deletion record in the structure (in a hashmap or rb-tree), wouldn't be called. Therefore, I propose running the `master` on its own separate benchmark so that it does practically all the same work except for the lookups and inserts into the `dels` structure. That is, the requirement is for the `master` to have the same number of fibers, the same number of secondary indexes, each fiber inserts the same number of tuples per transaction, and the length of each chain is the same. At the same time, all transactions must commit successfully.

To achieve this, we can, for example, have each fiber insert a set of `REPLACE_PER_TXN_COUNT` tuples, where all the tuples are different, not identical as before. The chain lengths will be the same (in terms of mathematical expectation), but all the replaces from a single fiber will land in different chains and will not cause conflicts.

<details><summary>
init.lua
</summary>

```lua
fiber = require('fiber')

local index_count = tonumber(arg[1])
local fiber_count = tonumber(arg[2])
local REPLACE_PER_TXN_COUNT = tonumber(arg[3]) or 2
local divisor = tonumber(arg[4]) or 3

os.execute('rm -rf *.snap *.xlog')

box.cfg{memtx_use_mvcc_engine=true}

box.schema.space.create("test")

for i = 1, index_count do
	box.space.test:create_index(
		('idx_%d'):format(i),
		{parts={{i}}, unique=true}
	)
end

local fun = require('fun')
local tuples = {}

for i = 1, fiber_count do
    tuples[i] = fun.totable(fun.take(index_count, fun.duplicate(i)))
end

local fibers = {}
local conds = {}
for i = 1, fiber_count do
	conds[i] = fiber.cond()
end
for i = 1, fiber_count do
	-- each chain has length REPLACE_PER_TXN_COUNT * X,
	-- where X ~ Bin(fiber_count, divisor/fiber_count)
	-- and E[X] = divisor and Var[X] = divisor(1 - divisor/fiber_count)
	local value = math.random(1, math.floor(fiber_count / divisor))
	local tuple = tuples[value] -- {value, value, ..., value}
	fibers[i] = fiber.create(function()
		box.begin()
			for _ = 1, REPLACE_PER_TXN_COUNT do
				box.space.test:replace(tuple)
			end
			conds[i]:wait()
		box.commit()
	end)
	fibers[i]:set_joinable(true)
end

for i = 1, fiber_count do
	conds[i]:signal()
	local ok, err = fibers[i]:join()
	assert(ok and err == nil)
end

os.exit()
```
</details>


<details><summary>
init_master.lua
</summary>

```lua
fiber = require('fiber')

local index_count = tonumber(arg[1])
local fiber_count = tonumber(arg[2])
local REPLACE_PER_TXN_COUNT = tonumber(arg[3]) or 2
local divisor = tonumber(arg[4]) or 3

os.execute('rm -rf *.snap *.xlog')

box.cfg{memtx_use_mvcc_engine=true}

box.schema.space.create("test")

for i = 1, index_count do
	box.space.test:create_index(
		('idx_%d'):format(i),
		{parts={{i}}, unique=true}
	)
end

local fun = require('fun')
local tuples = {}

for i = 1, fiber_count do
    tuples[i] = fun.totable(fun.take(index_count, fun.duplicate(i)))
end

local fibers = {}
local conds = {}
for i = 1, fiber_count do
	conds[i] = fiber.cond()
end

local max_value = math.floor(fiber_count / divisor)
assert(REPLACE_PER_TXN_COUNT <= max_value, 
       string.format("REPLACE_PER_TXN_COUNT (%d) must be <= max_value (%d)", 
                     REPLACE_PER_TXN_COUNT, max_value))
local values = fun.range(1, max_value):totable()

for i = 1, fiber_count do
	-- each chain has length REPLACE_PER_TXN_COUNT * X,
	-- where X ~ Bin(fiber_count, divisor/fiber_count)
	-- and E[X] = divisor and Var[X] = divisor(1 - divisor/fiber_count)
    local replaces = {}
    for j = 1, REPLACE_PER_TXN_COUNT do
        local value_pos = math.random(j, max_value)
        values[j], values[value_pos] = values[value_pos], values[j]
        replaces[j] = values[j]
    end

	fibers[i] = fiber.create(function()
		box.begin()
			for j = 1, REPLACE_PER_TXN_COUNT do
                local tuple = tuples[replaces[j]] -- {value, value, ..., value}
                box.space.test:replace(tuple)
			end
			conds[i]:wait()
		box.commit()
	end)
	fibers[i]:set_joinable(true)
end

for i = 1, fiber_count do
	conds[i]:signal()
	local ok, err = fibers[i]:join()
    assert(ok and err == nil)
end

os.exit()
```
</details>

<details><summary>
config
</summary>

```json
...
  "benchmark": {
    "benchmark_file": "init.lua",
    "arguments": {
      "index_count": 3,
      "fiber_count": {"start": 1000, "end": 10000, "points": 10},
      "REPLACE_PER_TXN_COUNT": 2,
      "divisor": 3
    },
    "argument_order": ["index_count", "fiber_count", "REPLACE_PER_TXN_COUNT", "divisor"],
    "test_iters": 10,
    "timeout": 60
  },
  "build_mode": "release",
  "versions": [
    {
      "branch": "master",
      "name": "current master",
      "benchmark_config": {
		// свой скрипт для мастера, который не вызывает конфликты
        "benchmark_file": "init_master.lua"
      }
	...
```
</details>

The length of each chain is approximately equal to `divisor * REPLACE_PER_TXN_COUNT = 6`.
Each test consisted of 10 runs, the results are averaged to reduce the variance.

<details><summary>
results :rocket:
</summary>

```
================================================================================================================
Benchmark Results
================================================================================================================

Version: current master
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         2                      3           0.118334                                     +0.00%
3            2000         2                      3           0.144787                                     +0.00%
3            3000         2                      3           0.173896                                     +0.00%
3            4000         2                      3           0.199382                                     +0.00%
3            5000         2                      3           0.233208                                     +0.00%
3            6000         2                      3           0.267084                                     +0.00%
3            7000         2                      3           0.293911                                     +0.00%
3            8000         2                      3           0.315084                                     +0.00%
3            9000         2                      3           0.348891                                     +0.00%
3            10000        2                      3           0.379855                                     +0.00%

Version: hashmap in each story
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         2                      3           0.128540                                     +8.63%
3            2000         2                      3           0.156785                                     +8.29%
3            3000         2                      3           0.186597                                     +7.30%
3            4000         2                      3           0.216743                                     +8.71%
3            5000         2                      3           0.244824                                     +4.98%
3            6000         2                      3           0.276520                                     +3.53%
3            7000         2                      3           0.302536                                     +2.93%
3            8000         2                      3           0.328043                                     +4.11%
3            9000         2                      3           0.364098                                     +4.36%
3            10000        2                      3           0.394441                                     +3.84%

Version: rb tree in each story
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         2                      3           0.112988                                     -4.52%
3            2000         2                      3           0.138321                                     -4.47%
3            3000         2                      3           0.168041                                     -3.37%
3            4000         2                      3           0.198170                                     -0.61%
3            5000         2                      3           0.230864                                     -1.01%
3            6000         2                      3           0.263849                                     -1.21%
3            7000         2                      3           0.290435                                     -1.18%
3            8000         2                      3           0.315639                                     +0.18%
3            9000         2                      3           0.346755                                     -0.61%
3            10000        2                      3           0.379768                                     -0.02%

Version: hashmap in tx manager
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         2                      3           0.110716                                     -6.44%
3            2000         2                      3           0.141590                                     -2.21%
3            3000         2                      3           0.170974                                     -1.68%
3            4000         2                      3           0.205375                                     +3.01%
3            5000         2                      3           0.231479                                     -0.74%
3            6000         2                      3           0.261592                                     -2.06%
3            7000         2                      3           0.286931                                     -2.37%
3            8000         2                      3           0.320998                                     +1.88%
3            9000         2                      3           0.347829                                     -0.30%
3            10000        2                      3           0.386947                                     +1.87%

================================================================================================================
```
</details>

### benchmark 2

Alternatively, we could use `init.lua` for all branches, simply setting `REPLACE_PER_TXN_COUNT = 1` to avoid conflicts on the master. This essentially tests everything we need: after each transaction commit, for all elements in the chain, there will be 1 delete + 1 insert + 1 lookup in the `dels` structure. The chains have a length of about 3.

<details><summary>
config
</summary>

```json
...
  "benchmark": {
    "benchmark_file": "init.lua",
    "arguments": {
      "index_count": 3,
      "fiber_count": {"start": 1000, "end": 10000, "points": 10},
      "REPLACE_PER_TXN_COUNT": 1,
      "divisor": 3
    },
    "argument_order": ["index_count", "fiber_count", "REPLACE_PER_TXN_COUNT", "divisor"],
    "test_iters": 10,
    "timeout": 60
  },
...
```
</details>


<details><summary>
results :rocket:
</summary>

```
================================================================================================================
Benchmark Results
================================================================================================================

Version: current master
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         1                      3           0.125039                                     +0.00%
3            2000         1                      3           0.150673                                     +0.00%
3            3000         1                      3           0.176661                                     +0.00%
3            4000         1                      3           0.205302                                     +0.00%
3            5000         1                      3           0.233695                                     +0.00%
3            6000         1                      3           0.258569                                     +0.00%
3            7000         1                      3           0.283362                                     +0.00%
3            8000         1                      3           0.311492                                     +0.00%
3            9000         1                      3           0.337357                                     +0.00%
3            10000        1                      3           0.364230                                     +0.00%

Version: hashmap in each story
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         1                      3           0.110677                                    -11.49%
3            2000         1                      3           0.138778                                     -7.89%
3            3000         1                      3           0.167491                                     -5.19%
3            4000         1                      3           0.198044                                     -3.54%
3            5000         1                      3           0.234689                                     +0.43%
3            6000         1                      3           0.258455                                     -0.04%
3            7000         1                      3           0.290934                                     +2.67%
3            8000         1                      3           0.323289                                     +3.79%
3            9000         1                      3           0.346990                                     +2.86%
3            10000        1                      3           0.370626                                     +1.76%

Version: rb tree in each story
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         1                      3           0.108522                                    -13.21%
3            2000         1                      3           0.136884                                     -9.15%
3            3000         1                      3           0.169787                                     -3.89%
3            4000         1                      3           0.194189                                     -5.41%
3            5000         1                      3           0.228883                                     -2.06%
3            6000         1                      3           0.259408                                     +0.32%
3            7000         1                      3           0.292401                                     +3.19%
3            8000         1                      3           0.316150                                     +1.50%
3            9000         1                      3           0.352273                                     +4.42%
3            10000        1                      3           0.369005                                     +1.31%

Version: hashmap in tx manager
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            1000         1                      3           0.114198                                     -8.67%
3            2000         1                      3           0.142020                                     -5.74%
3            3000         1                      3           0.172774                                     -2.20%
3            4000         1                      3           0.195575                                     -4.74%
3            5000         1                      3           0.229784                                     -1.67%
3            6000         1                      3           0.256391                                     -0.84%
3            7000         1                      3           0.285350                                     +0.70%
3            8000         1                      3           0.324344                                     +4.13%
3            9000         1                      3           0.348301                                     +3.24%
3            10000        1                      3           0.375519                                     +3.10%

================================================================================================================
```
</details>

### benchmark 3

Previously, we examined how the execution time depends on the number of fibers, with the average chain length (`divisor * REPLACE_PER_TXN_COUNT`) being fixed at 6 and 3 in the first and second benchmarks, respectively. This is much closer to real-world scenarios than long chains, as @drewdzzz rightly noted. However, let's fix `fiber_count`, for example, at 1000, and look at the dependency of execution time on the average chain length. In the results below, chain length = `divisor`, since `REPLACE_PER_TXN_COUNT = 1`.

<details><summary>
config
</summary>

```json
...
  "benchmark": {
    "benchmark_file": "init.lua",
    "arguments": {
      "index_count": 3,
      "fiber_count": 10000,
      "REPLACE_PER_TXN_COUNT": 1,
      "divisor": [6, 7, 8, 9, 10, 20, 50, 100, 1000, 10000]
    },
    "argument_order": ["index_count", "fiber_count", "REPLACE_PER_TXN_COUNT", "divisor"],
    "test_iters": 10,
    "timeout": 60
  },
...
```
</details>


<details><summary>
results :rocket:
</summary>

```
================================================================================================================
Benchmark Results
================================================================================================================

Version: current master
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            10000        1                      6           0.394508                                     +0.00%
3            10000        1                      7           0.397203                                     +0.00%
3            10000        1                      8           0.384789                                     +0.00%
3            10000        1                      9           0.378456                                     +0.00%
3            10000        1                      10          0.376238                                     +0.00%
3            10000        1                      20          0.384513                                     +0.00%
3            10000        1                      50          0.393643                                     +0.00%
3            10000        1                      100         0.425423                                     +0.00%
3            10000        1                      1000        0.919235                                     +0.00%
3            10000        1                      10000       4.471332                                     +0.00%

Version: hashmap in each story
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            10000        1                      6           0.384581                                     -2.52%
3            10000        1                      7           0.381149                                     -4.04%
3            10000        1                      8           0.392982                                     +2.13%
3            10000        1                      9           0.383030                                     +1.21%
3            10000        1                      10          0.385867                                     +2.56%
3            10000        1                      20          0.390776                                     +1.63%
3            10000        1                      50          0.435852                                    +10.72%
3            10000        1                      100         0.487083                                    +14.49%
3            10000        1                      1000        1.369146                                    +48.94%
3            10000        1                      10000       8.431483                                    +88.57%

Version: rb tree in each story
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            10000        1                      6           0.376846                                     -4.48%
3            10000        1                      7           0.389350                                     -1.98%
3            10000        1                      8           0.377031                                     -2.02%
3            10000        1                      9           0.395916                                     +4.61%
3            10000        1                      10          0.390701                                     +3.84%
3            10000        1                      20          0.400598                                     +4.18%
3            10000        1                      50          0.425085                                     +7.99%
3            10000        1                      100         0.481179                                    +13.11%
3            10000        1                      1000        1.866406                                   +103.04%
3            10000        1                      10000       12.764027                                  +185.46%

Version: hashmap in tx manager
----------------------------------------------------------------------------------------------------------------
index_count  fiber_count  REPLACE_PER_TXN_COUNT  divisor     Time per Iteration (s)     Percentage vs Baseline  
----------------------------------------------------------------------------------------------------------------
3            10000        1                      6           0.389861                                     -1.18%
3            10000        1                      7           0.387859                                     -2.35%
3            10000        1                      8           0.384880                                     +0.02%
3            10000        1                      9           0.390109                                     +3.08%
3            10000        1                      10          0.381513                                     +1.40%
3            10000        1                      20          0.390137                                     +1.46%
3            10000        1                      50          0.423088                                     +7.48%
3            10000        1                      100         0.468539                                    +10.13%
3            10000        1                      1000        1.145792                                    +24.65%
3            10000        1                      10000       6.677246                                    +49.33%

================================================================================================================
```
</details>


It can be seen that with a chain length of up to 50, the performance drop is no more than 10%, which seems quite acceptable as for me.


All the results above were obtained using a small, self-made utility: https://github.com/Astronomax/tarantool-bench. If you feel that a particular test case is missing here, you can run it yourself fairly quickly. All you need to do is write a .lua script, make some minor edits to the .json configuration file in the repository, and then simply run `python3 benchmark_runner.py benchmark_config.json --tarantool-path <tarantool dir path>`. It will automatically build each branch in a separate directory, run the benchmarks, and dump the results.

Astronomax committed:
memtx: reduce false-positive aborts in Memtx-MVCC

It was discovered that Memtx MVCC aborts transactions too frequently, even
in cases where it would seem possible to avoid doing so. The issue was in
how potential duplicates in secondary indexes were handled when a statement
became "prepared".

The processing worked as follows. When a replace/insert statement `stmt`
becomes "prepared" (`stmt->add_story != NULL`):

- Iterate through each chain (except the primary one) upwards from
`stmt->add_story`
- For each `story` along the path, check:
    - If the `story` was added by the same transaction as
      `stmt->add_story`, skip it. The check has already been performed
      for such stories and nothing could have been corrupted.
    - If the `story` was added by a transaction that previously deleted
      (**using DELETE statement**) a tuple with the same key, skip it. It
      couldn't have been a duplicate of `stmt->add_story->tuple` because
      at the serialization point when `story->tuple` was inserted,
      `stmt->add_story->tuple` was definitely deleted.
    - If `story->tuple` has a primary key different from the primary key
      of `stmt->add_story->tuple`
      (`story->add_stmt->del_story != stmt->add_story`), then abort
      `story->add_stmt->txn` - it's a duplicate.

This patch addresses the second point. Actually, it doesn't matter to us
which type of statement deleted a tuple with the same key: a `DELETE` or
a `REPLACE`. Up until now, we have only handled `DELETE`, because its
`del_story` reference is constant and always points to the same tuple,
and the read tracker "monitors" the lifetime of the tuple pointed to by
this reference. Because the reference is set only once, it was easy for
us to handle this case. With `REPLACE` statements, things are a bit more
complicated, as their `del_story` reference can change any number of
times. At different points in time, this property ("the story was added by
a transaction that previously replaced a tuple with the same key") can
toggle between true and false as other transactions become "prepared".
Therefore, we need some mechanism to quickly handle the queries
"transaction `TXN` has become prepared" and "does our property hold for
story `story`?".

Closes #11927, #11942

NO_DOC=enhancement

drewdzzz commented:
Discussed f2f. Approach with flag looks very nice to me because it doesn't slow down MVCC, but it complicates its logic. We should try to work it out to make it more straightforward and convenient.

Also, this issue is not really critical, so we decided to prioritize other more important issues over this one.
